src/Cli/dotnet/commands/dotnet-test/TestApplicationActionQueue.cs

```diff
@@ -36,13 +40,18 @@ public void EnqueueCompleted()
             _channel.Writer.Complete();
         }
 
-        private async Task Read(Func<TestApplication, Task> action)
+
+        private async Task Read(Func<TestApplication, Task<int>> action)
         {
             while (await _channel.Reader.WaitToReadAsync())
             {
                 if (_channel.Reader.TryRead(out TestApplication testApp))
                 {
-                    await action(testApp);
+                    int result = await action(testApp);
+                    lock (_lock)
+                    {
+                        _hasFailed |= result != ExitCodes.Success;
```

Comment:
I would remove the lock and simply check if result is not zero and set to true. Here the race is benign, if we will enter in the if we will go to set to true..so the race could read a current _hasFailed true or false...but anyway we will set to true at the end, so the state will be ok.